### PR TITLE
add which kwarg to autofmt_xdate

### PR DESCRIPTION
**Pull request by patniharshit**
refs #8128 
Review required.
Also help needed in writing the test. Firstly I am not sure where to write the test, because a quick search in test_figure.py did not return any tests related to autofmt_xdate.

### CODE REVIEW

File: lib/matplotlib/figure.py [hunk 1]
```diff
@@ -448,29 +448,35 @@ def autofmt_xdate(self, bottom=0.2, rotation=30, ha='right'):
...
-        *rotation*
+        rotation : angle in degrees
             The rotation of the xtick labels
 
-        *ha*
+        ha : string
             The horizontal alignment of the xticklabels
+
+        which : {None, 'major', 'minor', 'both'}
+            Selects which ticklabels to return
```
Comment (tacaswell) [inline lib/matplotlib/figure.py]: 'which ticks to rotate'?

**Comment by tacaswell** [comment]
If there are no tests, then you are in luck as you get to meaningfully increase our test coverage!

It is a `figure` method so putting it in `test_figure` would make sense, but it is also date-specific so putting in `test_dates` would also make sense.   I lean slightly towards the second, but does not matter that much.

**Comment by patniharshit** [comment]
No tests found in test_dates as well.

File: lib/matplotlib/figure.py (hunk 1 above)
Comment (patniharshit) [inline lib/matplotlib/figure.py]: What 'which' is referring to is not clear or is this a suggestion to change return to rotate?

Comment (phobson) [inline lib/matplotlib/figure.py]: Tom meant that you should change "Selects which ticklabels to return" to "Selects which ticklabels to rotate"

Comment (patniharshit) [inline lib/matplotlib/figure.py]: Done.

**Comment by NelleV** [comment]
I agree with @tacaswell that a test is needed for this new feature.

File: lib/matplotlib/tests/test_figure.py [hunk 2]
```diff
@@ -234,3 +236,38 @@ def test_figaspect():
...
+        '13 Jan 2013', '14 Jan 2013']
+
+        time = ['16:44:00', '16:45:00', '16:46:00', '16:47:00', '16:48:00', '16:49:00',
+        '16:51:00', '16:52:00', '16:53:00', '16:55:00', '16:56:00', '16:57:00']
+
+        angle = 60
+        minors = [1,2,3,4,5,6,7]
+        
+        x = mdates.datestr2num(date)
+	y = mdates.datestr2num(time)
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_figure.py]: It looks like you have a mix of tabs and spaces, can you please make them all spaces (many editors have tools to help with this, and a minimum find and replace `\t' -> '    '.

**Comment by tacaswell** [comment]
Can you also drop a file in https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new so that this gets properly advertised?

File: lib/matplotlib/figure.py [hunk 3]
```diff
@@ -448,29 +448,35 @@ def autofmt_xdate(self, bottom=0.2, rotation=30, ha='right'):
...
+        ha : string
             The horizontal alignment of the xticklabels
+
+        which : {None, 'major', 'minor', 'both'}
+            Selects which ticklabels to rotate
         """
         allsubplots = all(hasattr(ax, 'is_last_row') for ax in self.axes)
         if len(self.axes) == 1:
-            for label in self.axes[0].get_xticklabels():
+            for label in self.axes[0].get_xticklabels(False, which):
```
Comment (tacaswell) [inline lib/matplotlib/figure.py]: You can also pass this in as `get_xticklabels(which=which)` which is less cryptic than the leading `False`.

Which we support _both_ `minor` and `which` is an interesting question...which the answer to is it is my fault: 08d961a176c3376571f63fb3b7ae01e56137cde7 😞 .

**Comment by patniharshit** [comment]
whats the problem here? PEP8?

**Comment by phobson** [comment]
@patniharshit yeah you've got a number of PEP8 violations. 
https://travis-ci.org/matplotlib/matplotlib/jobs/210002135#L1691

**Comment by patniharshit** [comment]
"AppVeyor build failed" what does this mean?

**Comment by phobson** [comment]
@patniharshit Travis CI tests the code on a linux operating system. AppVeyor tests the code on Windows. However, AppVeyor (the system) is generally broken for us right now, so we're not too concerned that the build failed.

**Comment by patniharshit** [comment]
What do you mean by parameterizing the tests?

**Comment by phobson** [comment]
@patniharshit 

In this case, I would take away the `for` loop as pass those values in to the test function itself. This has the advantage of testing every value of `which` independently. That means is that is `None` fails for some reason, the other values still get tested. In your case, that means your test would become:

```python
@pytest.mark.parametrize('which', [None, 'both', 'major', 'minor'])
def test_autofmt_xdate(which):
     date = ['3 Jan 2013', '4 Jan 2013', '5 Ja....
     ...
```

File: lib/matplotlib/tests/test_figure.py [hunk 4]
```diff
@@ -234,3 +237,38 @@ def test_figaspect():
     assert h / w == 0.5
     w, h = plt.figaspect(np.zeros((2, 2)))
     assert h / w == 1
+
+@pytest.mark.parametrize('which', [None, 'both', 'major', 'minor'])
```
Comment (phobson) [inline lib/matplotlib/tests/test_figure.py]: pep8 will want another blank line here

**Comment by patniharshit** [comment]
Can someone help me here? Why Travis is failing? All tests are passing locally.

**Comment by phobson** [comment]
Travis failure looks like it's just a flaky test unrelated to your PR. Looks like the Appveyor build is old enough that it was suffering from systematic failures.

File: doc/users/whats_new/autofmt_xdate.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+The autofmt_xdate of the figure.Figure class previously formatted only major tick labels but now formats both minor and major tick labels by passing the argument "which = {None, 'minor', 'major', 'both'}".
```
Comment (tacaswell) [inline doc/users/whats_new/autofmt_xdate.rst]: Can you please use the template in `README.rst`?  That makes in very easy to combine each of these files when it comes time for a release.

**Comment by tacaswell** [comment]
:+1:  except for the change to the docs.

File: doc/users/whats_new/autofmt_xdate.rst (hunk 5 above)
Comment (patniharshit) [inline doc/users/whats_new/autofmt_xdate.rst]: Didn't notice there was a template. Updated now.

File: lib/matplotlib/tests/test_figure.py [hunk 6]
```diff
@@ -234,3 +237,39 @@ def test_figaspect():
     assert h / w == 0.5
     w, h = plt.figaspect(np.zeros((2, 2)))
     assert h / w == 1
+
+
+@pytest.mark.parametrize('which', [None, 'both', 'major', 'minor'])
+def test_autofmt_xdate(which):
+    date = ['3 Jan 2013', '4 Jan 2013', '5 Jan 2013', '6 Jan 2013',
+    '7 Jan 2013', '8 Jan 2013', '9 Jan 2013', '10 Jan 2013', '11 Jan 2013',
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py]: Align with the opening bracket.

File: lib/matplotlib/tests/test_figure.py [hunk 7]
```diff
@@ -234,3 +237,39 @@ def test_figaspect():
...
+    '12 Jan 2013', '13 Jan 2013', '14 Jan 2013']
+
+    time = ['16:44:00', '16:45:00', '16:46:00', '16:47:00', '16:48:00',
+    '16:49:00', '16:51:00', '16:52:00', '16:53:00', '16:55:00', '16:56:00',
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py]: Here too.

File: lib/matplotlib/tests/test_figure.py [hunk 8]
```diff
@@ -234,3 +237,39 @@ def test_figaspect():
...
+    ax.plot(x, y)
+    ax.yaxis_date()
+    ax.xaxis_date()
+
+    ax.xaxis.set_minor_locator(AutoMinorLocator(2))
+    ax.xaxis.set_minor_formatter(FixedFormatter(minors))
+
+    fig.autofmt_xdate(0.2, angle, 'right', which)
+
+    if((which == 'both') or (which == 'major') or (which is None)):
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py]: `if which in ('both', 'major', None):`

File: lib/matplotlib/tests/test_figure.py [hunk 9]
```diff
@@ -234,3 +237,39 @@ def test_figaspect():
...
+        for label in fig.axes[0].get_xticklabels(False, 'major'):
+            assert (int(label.get_rotation()) == angle)
+
+    if(which == 'both' or which == 'minor'):
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py]: `if which in ('both', 'minor'):`

File: lib/matplotlib/tests/test_figure.py [hunk 10]
```diff
@@ -234,3 +237,39 @@ def test_figaspect():
...
+    ax.xaxis_date()
+
+    ax.xaxis.set_minor_locator(AutoMinorLocator(2))
+    ax.xaxis.set_minor_formatter(FixedFormatter(minors))
+
+    fig.autofmt_xdate(0.2, angle, 'right', which)
+
+    if((which == 'both') or (which == 'major') or (which is None)):
+        for label in fig.axes[0].get_xticklabels(False, 'major'):
+            assert (int(label.get_rotation()) == angle)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py]: Unnecessary parentheses.

File: lib/matplotlib/tests/test_figure.py [hunk 11]
```diff
@@ -234,3 +237,39 @@ def test_figaspect():
...
+
+    if(which == 'both' or which == 'minor'):
+        for label in fig.axes[0].get_xticklabels(True, 'minor'):
+            assert (int(label.get_rotation()) == angle)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_figure.py]: Unnecessary parentheses.

File: lib/matplotlib/figure.py [hunk 12]
```diff
@@ -1838,7 +1844,6 @@ def ginput(self, n=1, timeout=30, show_clicks=True, mouse_add=1,
 
         Right clicking cancels last input.
 
-        The buttons used for the various actions (adding points, removing
```
Comment (QuLogic) [inline lib/matplotlib/figure.py]: Why is this removed?

File: lib/matplotlib/figure.py [hunk 13]
```diff
@@ -448,29 +448,35 @@ def autofmt_xdate(self, bottom=0.2, rotation=30, ha='right'):
...
-        *rotation*
+        rotation : angle in degrees
             The rotation of the xtick labels
 
-        *ha*
+        ha : string
             The horizontal alignment of the xticklabels
+
+        which : {None, 'major', 'minor', 'both'}
+            Selects which ticklabels to rotate
```
Comment (QuLogic) [inline lib/matplotlib/figure.py]: This should indicate what the default (`None`) means; both, neither?

File: lib/matplotlib/figure.py (hunk 12 above)
Comment (patniharshit) [inline lib/matplotlib/figure.py]: Thanks for pointing that out, it was probably a mistake on my part.

File: lib/matplotlib/figure.py (hunk 13 above)
Comment (patniharshit) [inline lib/matplotlib/figure.py]: Updated.

File: lib/matplotlib/tests/test_figure.py (hunk 11 above)
Comment (patniharshit) [inline lib/matplotlib/tests/test_figure.py]: Removed.

File: lib/matplotlib/tests/test_figure.py (hunk 9 above)
Comment (patniharshit) [inline lib/matplotlib/tests/test_figure.py]: Removed.

Comment (patniharshit) [inline lib/matplotlib/tests/test_figure.py]: Changed.

File: lib/matplotlib/tests/test_figure.py (hunk 10 above)
Comment (patniharshit) [inline lib/matplotlib/tests/test_figure.py]: Removed.

File: lib/matplotlib/tests/test_figure.py (hunk 8 above)
Comment (patniharshit) [inline lib/matplotlib/tests/test_figure.py]: Changed.

File: lib/matplotlib/tests/test_figure.py (hunk 7 above)
Comment (patniharshit) [inline lib/matplotlib/tests/test_figure.py]: Done.

File: lib/matplotlib/tests/test_figure.py (hunk 6 above)
Comment (patniharshit) [inline lib/matplotlib/tests/test_figure.py]: Done.

File: doc/users/whats_new/autofmt_xdate.rst (hunk 5 above)
Comment (patniharshit) [inline doc/users/whats_new/autofmt_xdate.rst]: Yup, forgot to delete that. Deleted now.

File: lib/matplotlib/tests/test_figure.py [hunk 14]
```diff
@@ -234,3 +237,39 @@ def test_figaspect():
     assert h / w == 0.5
     w, h = plt.figaspect(np.zeros((2, 2)))
     assert h / w == 1
+
+
+@pytest.mark.parametrize('which', [None, 'both', 'major', 'minor'])
```
Comment (anntzer) [inline lib/matplotlib/tests/test_figure.py]: At the risk of sounding pedantic I'll note that this test belongs to test_dates.py.

**Comment by phobson** [comment]
@patniharshit thanks for sticking with us! I fixed a minor pep8 thing. I'll wait for the CI to wrap up on that commit and then I'll merge.

**Comment by phobson** [comment]
@patniharshit thanks again -- merged!
